### PR TITLE
🎉 Source Hubspot: Migrate to CDK

### LINKED ISSUE #5114: Source Hubspot: migrate to CDK
## Tell us about the problem you're trying to solve
In order to keep the code consistent across the connectors and to be able to test this connector the same way we do for the rest we need to move it to the new base.

## Describe the solution you’d like
Remove all of the `base-python` and `airbyte-protocol` references and replace then with `airbyte-cdk`.  Refactor the codebase to fit the current best practices. See https://docs.airbyte.io/connector-development/cdk-python, https://docs.airbyte.io/connector-development/best-practices and other `airbyte-cdk` sources for the refs (such as `Harvest`, `Braintree` etc.).

## Describe the alternative you’ve considered or used
We may just left the connector as it for now. But still at some point we'll need to rebuild this connector.

## Additional context
Description updated after the https://github.com/airbytehq/airbyte/pull/6465.

## Are you willing to submit a PR?

I can do it.

### PR DESCRIPTION
**Pull request by augan-rymkhan**
## What
Resolves [#5114](https://github.com/airbytehq/airbyte/issues/5114)
Refactor the connector to use CDK classes.

## How
**Make SourceHubspot inherits from AbstractSource**
- [x] Implement `check_connection`, we can use the existing logic prodived by [client.health_check](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-hubspot/source_hubspot/client.py#L115) in this method
- [x] Fix `stream.name` property so  `discover` command can find the related schemas and add dynamic properties by overriding `get_json_schema` method. To do that refactor stream class names. Schema names should not be changed to support backward compatibilty.
- [x] Implement `streams` method, move logic in `Client._apis` to this method and change the its type from dict to list.
- [x] Remove `Client` class.

**Refactor Stream classes**
- [x]  Make Stream base class inherit from `airbyte_cdk.sources.streams.core import Stream`
- [x] Mark `IncrementalStream` stream as support incremental by adding field `cursor_field`
- [x] Add method  `get_updated_state` to `IncrementalStream` class, which returns `self.state or {self.state_pk: self._start_date}`
IMO, we should not change `state` property because it has complex logic of handling stream state.
- [x] Add `read_records` method to Stream class, which runs generator `list_records(fields=fields)`
- [x] Override `read_records` method in `IncrementalStream` so it sets initial state got from connector state:
 ```
        if stream_state:
            self.state = stream_state
        yield from self.list_records(fields=self.get_fields())
```

## Recommended reading order
1. `source-hubspot/source_hubspot/source.py`
2. `source-hubspot/source_hubspot/api.py`
3. `source-hubspot/source_hubspot/client.py`
4. `source-hubspot/source_hubspot/test_source.py`
5. `source-hubspot/acceptance-test-config.yml`

### CODE REVIEW

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py [hunk 1]
```diff
@@ -199,6 +201,8 @@ class Stream(ABC):
     limit_field = "limit"
     limit = 100
     offset = 0
+    _name: str = None
```
Comment (augan-rymkhan) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:204]: This property is added to define stream name when use CRMSearchStream, CRMObjectIncrementalStream general (template) stream. So self._name = "companies"  matches to  `schema/companies.json`

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py [hunk 2]
```diff
@@ -211,10 +215,28 @@ def __init__(self, api: API, start_date: str = None, **kwargs):
...
-            stream_name = stream_name[: -len("Stream")]
-        return stream_name
+        if self._name:
+            return self._name
+        return super().name
+
+    def get_json_schema(self) -> Mapping[str, Any]:
+        json_schema = super().get_json_schema()
+        if self.properties:
+            json_schema["properties"]["properties"] = {"type": "object", "properties": self.properties}
```
Comment (augan-rymkhan) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:225]: Some entities has dynamic set of properties, so we trying to resolve those at runtime

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py [hunk 3]
```diff
@@ -467,12 +489,37 @@ class IncrementalStream(Stream, ABC):
...
 
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ):
+        if stream_state:
+            self.state = stream_state
```
Comment (augan-rymkhan) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:511]: Set initial `state` which is got from `connector_state` mapping, passed from `AbstractSource` `_read_incremental` method.

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py [hunk 4]
```diff
@@ -662,8 +710,12 @@ class CRMObjectIncrementalStream(CRMObjectStream, IncrementalStream):
     limit = 100
     need_chunk = False
 
+    def __init__(self, name: str = None, **kwargs):
+        super().__init__(**kwargs)
+        self._name = name
+
 
-class CampaignStream(Stream):
+class Campaigns(Stream):
```
Comment (augan-rymkhan) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:718]: Refactored stream names so its snake case version matches to `schemas/campaigns.json`

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/source.py [hunk 5]
```diff
@@ -2,25 +2,202 @@
...
+        # TODO assert all streams exist in the connector
+        # get the streams once in case the connector needs to make any queries to generate them
+        stream_instances = {s.name: s for s in self.streams(config)}
+        self._stream_to_instance_map = stream_instances
+        with create_timer(self.name) as timer:
+            for configured_stream in catalog.streams:
+                stream_instance = stream_instances.get(configured_stream.stream.name)
+                if not stream_instance and configured_stream.stream.name == "quotes":
+                    logger.warning("Stream `quotes` does not exist in the source. Skip reading `quotes` stream.")
+                    continue
```
Comment (augan-rymkhan) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/source.py:158]: This method is overridden to check whether the stream `quotes` exists in the source, if not skip reading that stream.

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/source.py [hunk 6]
```diff
@@ -2,25 +2,202 @@
...
+        api = self.get_api(config=config)
+        common_params = dict(api=api, start_date=start_date)
+        return common_params
+
+    def streams(self, config: Mapping[str, Any]) -> List[Stream]:
+        credentials = config.get("credentials", {})
+        common_params = self.get_common_params(config=config)
+        streams = [
+            Campaigns(**common_params),
+            CRMSearchStream(
```
Comment (keu) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/source.py:71]: @augan-rymkhan I propose to follow here the practice we have in many other connectors:
make these attribute of class and create explicit classes for each stream, i.e.
```python
class Contacts(HubspotCRMSearchStream):
   entity = "contact"
   cursor_field = "lastmodifieddate"
   associations=["contacts"]
```

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py [hunk 7]
```diff
@@ -467,12 +489,37 @@ class IncrementalStream(Stream, ABC):
     # Flag which enable/disable chunked read in read_chunked method
     # False -> chunk size is max (only one slice), True -> chunk_size is 30 days
     need_chunk = True
+    state_checkpoint_interval = 500
+
+    @property
+    def cursor_field(self) -> Union[str, List[str]]:
+        return self.updated_at_field
```
Comment (keu) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:496]: unnecessary indirection, lets just rename it to cursor_field

**Review by keu — CHANGES_REQUESTED** [review]
nice job! but the current PR feels more like a wrapper/adapter around the legacy codebase to fit cdk interface,
not sure if that matches the description of the ticket - i.e. the code doesn't look like following our best practice and utilize CDK classes (slices, HTTPStream, etc)
just a sidenote

**Review by keu — APPROVED** [review]
lgtm

**Review by sergei-solonitcyn — APPROVED** [review]

File: airbyte-integrations/connectors/source-hubspot/acceptance-test-config.yml [hunk 8]
```diff
@@ -34,8 +34,8 @@ tests:
       future_state_path: "integration_tests/abnormal_state.json"
       cursor_paths:
         subscription_changes: ["timestamp"]
-        email_events: ["timestamp"]
-        contact_lists: ["timestamp"]
+        email_events: ["created"]
```
Comment (keu) [inline airbyte-integrations/connectors/source-hubspot/acceptance-test-config.yml:37]: if it matches the cursor_field in a stream (catalog) then you don't need cusor_paths for this stream (and any of them)
the purpose of `cursor_paths` is to handle cases when the path in the STATE file is nested
```json
"some_field": {
   "cursor_field": "value"
}
```

**Review by keu — APPROVED** [review]

**Review by sergei-solonitcyn — APPROVED** [review]

**Pull request merged**